### ISSUE TITLE
When using Teamviewer and MinTTY, Alt gr characters can't be typed

### ISSUE THREAD

**Issue by danowar2k** [mintty/mintty#783]
This seems to be a problem at least in MinTTY 2.8.5 (used in the current git for Windows 2.18.0).

When I use Alt gr to type {,},~, etc. in a MinTTY terminal while connected via Teamviewer (Guest and Host version: 13.1.3629), the characters do not appear.

MinTTY Host OS: Windows 7
Teamviewer Guest OS: Windows 7

Using the on screen keyboard of Windows is a workaround, it seems.

**Comment by mintty** [comment]
Does it work with other versions of mintty and/or teamviewer?
Does it work with another application or terminal?

**Comment by mintty** [comment]
Also, can this be tested on a single machine, i.e. TeamViewer connecting to itself? When I try that, it does nothing.

**Comment by danowar2k** [comment]
When I use Teamviewer and use other applications on the machine, Alt gr special characters work (e.g. the Windows command prompt or an SSH terminal inside IntelliJ).

I don't know if I can do a parallel install of lower Teamviewer versions, so I can't check right now if lower versions work.
The used Teamviewer version is the highest available.

I'll check if using a minTTY 2.9.0 terminal works...Do I need to check if lower versions than 2.8.5 work? If so, which ones?

I don't think Teamviewer can connect to the same machine. You need two Windows machines to validate this, a guest and a host.

**Comment by danowar2k** [comment]
Ok, I am a bit confused. I can find no installation guide. I guess I have to build minTTY 2.9.0 from source? Or are there any downloadable Windows binaries?

**Comment by mintty** [comment]
Change log 2.8.0:
  * Workaround for broken AltGr of Windows on-screen keyboard (#692).
So if you can test any version before 2.8.0 that might filter out this as a potential cause, for a start.

**Comment by mintty** [comment]
Building mintty from source for Git for Windows appears to be tricky because there is no development environment by default. Can you test with Cygwin?

**Comment by danowar2k** [comment]
Wouldn't testing with Cygwin change more than one "variable" in finding the bug? Meaning I use Cygwin instead of whatever git for Windows uses AND change the used MinTTY?

Of course I could first install Cygwin and use the same MinTTY, 2.8.5 to check...What do you think?

**Comment by danowar2k** [comment]
Okay, I've installed Cygwin. First, I installed mintty 2.8.5, and I can't use Alt gr special characters.

Then I upgraded to mintty 2.9.0. When I tried to do a }, it first worked but then I deleted the characters and it didn't work anymore. When I now enter a }, instead I get "7p". When I enter a ~, I get "7k".

Any more characters I should try?

Now I have the problem that I can only install mintty 2.8.4, 2.8.5 or 2.9.0 from the Cygwin setup GUI. I'll try to downgrade via CLI.

**Comment by danowar2k** [comment]
I can't seem to find any Cygwin package mirror providing anything but 2.8.4, 2.8.5 and 2.9.0...

**Comment by danowar2k** [comment]
Okay, now I'm confused. I tried to use the 2.7.9 version of mintty I found at wsltty with my git for Windows 2.18.0 . I replaced the binary and tried to start a git bash. I got the error that a cygwin.dll would be needed.

Okay, that error could have been expected, as git for Windows uses Msys and mingw64 and not cygwin.

I then looked for older git for windows files to get a correct compiled binary of mintty 2.7.9 .
I found one at https://github.com/git-for-windows/git-sdk-64/blob/29c8b9aad70742fd305da07056bb76cf454942af/usr/bin/mintty.exe

This should be a mintty 2.7.9 that was once used in git for Windows.
I replaced my git for Windows mintty with it, and started a new git bash. This time it worked.
I did a `mintty -V` and saw it was 2.7.9.

BUT this 2.7.9 version seems to have the bug that the cygwin 2.7.9 version doesn't have. I couldn't use Alt gr with it.

Now what?

**Comment by danowar2k** [comment]
mintty 2.6.2 via the same source also has the bug.

So maybe the MSYS mintty version is the problem?

**Comment by mintty** [comment]
To compiler older versions, you can download the source archive from https://github.com/mintty/mintty/releases or from https://github.com/mintty/release/. Then `make CCOPT=-Wno-error` to compensate for more picky newer gcc versions.

> Now what?

Good question. You said the problem occurs with the latest cygwin version, let's focus on that.
Please activate `#define debug_virtual_key_codes` in wininput.c (around line 1000) and recompile.
Then run mintty from another command line (e.g. also mintty) to make the debug output visible.
What does AltGr+~ etc produce?

**Comment by danowar2k** [comment]
Okay, I've installed the following cygwin packages:
```
Install libisl13 0.14.1-1 (automatically added)
Install cygwin32 2.10.0-1 (automatically added)
Install cygwin32-binutils 2.29-1 (automatically added)
Install cygwin32-default-manifest 6.4-1 (automatically added)
Install libmpc3 1.1.0-1 (automatically added)
Install libmpfr4 3.1.6-1p1 (automatically added)
Install cygwin32-w32api-headers 4.0.4-1 (automatically added)
Install cygwin32-w32api-runtime 4.0.4-1 (automatically added)
Install cygwin32-gcc-core 6.4.0-1
```
I hope those are the correct packages to use gcc

**Comment by danowar2k** [comment]
*sigh*

So I had to install the wget and make packages. Then I downloaded the sources for mintty 2.9.0 .

I replaced
`#define dont_debug_virtual_key_codes`
with
`#define debug_virtual_key_codes`

and used the make command you posted above.

Result:
```
$ make CCOPT=-Wno-error
#cd src; make exe
#cd src; make bin
cd src; make
make[1]: Verzeichnis „/home/poggenpohlda/mintty-2.9.0/src“ wird betreten
mkdir -p ../bin/cygwin64
gcc -c -MMD -MP -DTARGET=x86_64-pc-cygwin -DNDEBUG -std=gnu99 -include std.h -Wall -Wextra -Wundef -Werror -mtune=atom -fomit-frame-pointer -O2 -Wno-error -fstack-check windialog.c -o ../bin/cygwin64/windialog.o
gcc -c -MMD -MP -DTARGET=x86_64-pc-cygwin -DNDEBUG -std=gnu99 -include std.h -Wall -Wextra -Wundef -Werror -mtune=atom -fomit-frame-pointer -O2 -Wno-error -fstack-check std.c -o ../bin/cygwin64/std.o
gcc -c -MMD -MP -DTARGET=x86_64-pc-cygwin -DNDEBUG -std=gnu99 -include std.h -Wall -Wextra -Wundef -Werror -mtune=atom -fomit-frame-pointer -O2 -Wno-error -fstack-check minibidi.c -o ../bin/cygwin64/minibidi.o
gcc -c -MMD -MP -DTARGET=x86_64-pc-cygwin -DNDEBUG -std=gnu99 -include std.h -Wall -Wextra -Wundef -Werror -mtune=atom -fomit-frame-pointer -O2 -Wno-error -fstack-check mcwidth.c -o ../bin/cygwin64/mcwidth.o
gcc -c -MMD -MP -DTARGET=x86_64-pc-cygwin -DNDEBUG -std=gnu99 -include std.h -Wall -Wextra -Wundef -Werror -mtune=atom -fomit-frame-pointer -O2 -Wno-error -fstack-check winimg.c -o ../bin/cygwin64/winimg.o
gcc -c -MMD -MP -DTARGET=x86_64-pc-cygwin -DNDEBUG -std=gnu99 -include std.h -Wall -Wextra -Wundef -Werror -mtune=atom -fomit-frame-pointer -O2 -Wno-error -fstack-check wininput.c -o ../bin/cygwin64/wininput.o
wininput.c:998:10: schwerwiegender Fehler: _vk.t: No such file or directory
 #include "_vk.t"
          ^~~~~~~
Kompilierung beendet.
make[1]: *** [Makefile:139: ../bin/cygwin64/wininput.o] Fehler 1
make[1]: Verzeichnis „/home/poggenpohlda/mintty-2.9.0/src“ wird verlassen
make: *** [Makefile:26: exe] Fehler 2

```
So, what next?

**Comment by danowar2k** [comment]
Okay, I found it. The _vk.t target is in the src folder Makefile. I could make that. You probably have to change the main Makefile so that the _vk.t target is built at the right time.... I think...

Anyway, mintty 2.9.0 is now built.

**Comment by danowar2k** [comment]
Okay, I'll post typing results here.

compiled mintty 2.9.0
Keys: Alt gr + ~
Result: 7k
Logged:
```
win_key_down 0011 VK_CONTROL scan 29
win_key_down 0012 VK_MENU scan 312
win_key_down 00BB VK_OEM_PLUS scan 27
  win_key_up 00BB VK_OEM_PLUS
  win_key_up 0011 VK_CONTROL
  win_key_up 0011 VK_CONTROL
  win_key_up 0012 VK_MENU
```

Keys: Alt gr + {
Result: NOTHING
Logged:
```
win_key_down 0011 VK_CONTROL scan 29
win_key_down 0012 VK_MENU scan 312
win_key_down 0037 37 scan 8
  win_key_up 0037 37
  win_key_up 0011 VK_CONTROL
  win_key_up 0011 VK_CONTROL
  win_key_up 0012 VK_MENU
```

More data necessary?

**Comment by mintty** [comment]
Hmm, that looks quite like I see here, except that you have `win_key_up 0011 VK_CONTROL` doubled, but that should not matter. Please also enable `#define debug_key`.

**Comment by mintty** [comment]
Also, add this below `GetKeyboardState(kbd);` and function `inline bool is_key_down...`:
```
#ifdef debug_virtual_key_codes
  printf(" Shift %d:%d/%d Ctrl %d:%d/%d Alt %d:%d/%d\n",
         is_key_down(VK_SHIFT), is_key_down(VK_LSHIFT), is_key_down(VK_RSHIFT),
         is_key_down(VK_CONTROL), is_key_down(VK_LCONTROL), is_key_down(VK_RCONTROL),
         is_key_down(VK_MENU), is_key_down(VK_LMENU), is_key_down(VK_RMENU));
#endif
```

**Comment by danowar2k** [comment]
I've enabled debug_key and put the code ONCE below the two lines:
```
  GetKeyboardState(kbd);
  inline bool is_key_down(uchar vk) { return kbd[vk] & 0x80; }
#ifdef debug_virtual_key_codes
  printf(" Shift %d:%d/%d Ctrl %d:%d/%d Alt %d:%d/%d\n",
         is_key_down(VK_SHIFT), is_key_down(VK_LSHIFT), is_key_down(VK_RSHIFT),
         is_key_down(VK_CONTROL), is_key_down(VK_LCONTROL), is_key_down(VK_RCONTROL),
         is_key_down(VK_MENU), is_key_down(VK_LMENU), is_key_down(VK_RMENU));
#endif

```
I hope that is what you requested.

**Comment by danowar2k** [comment]
After testing all other shift, alt and control keys, I don't understand why Alt gr produces VK_CONTROL and VK_MENU, while Alt produces only VK_MENU.

**Comment by danowar2k** [comment]
Tried the following:

Keys: Ctrl + Alt + {
Result: NOTHING
Logged:
```
win_key_down 0012 VK_MENU scan 56
 Shift 0:0/0 Ctrl 0:0/0 Alt 1:1/0
win_key_down 0011 VK_CONTROL scan 29
 Shift 0:0/0 Ctrl 1:1/0 Alt 1:1/0
win_key_down 0037 37 scan 8
 Shift 0:0/0 Ctrl 1:1/0 Alt 1:1/0
undead 0037 scn 8 -> 1 0037
undead 0037 scn 8 -> 1 002F
 <-0... ctrl_key
  win_key_up 0037 37
  win_key_up 0012 VK_MENU
  win_key_up 0011 VK_CONTROL
```

**Comment by mintty** [comment]
Thanks so far. Unfortunately this doesn't yet give a clue what could be done. With the version you tested, does AltGr+~ not enter anything anymore?
The root cause is that Microsoft in their universal weirdness of API design did not spend a distinct VK code for AltGr. Instead, AltGr needs to be recognized by a combination of left-control and right-alt (or sometimes left-alt) and, to distinguish from true Ctrl+Alt combinations, even timing conditions. And even Microsoft's own onscreen keyboard is inconsistent (using left-alt rather than right-alt) which was the reason for the workaround in 2.8.0.
So the following questions might give us more diagnostics:
* In those applications where AltGr+~ works (i.e. enters ~), does Ctrl+Alt+~ make a difference?
* If you add timestamps to debug output: `  printf(" [%d] Shift %d:%d/%d Ctrl %d:%d/%d Alt %d:%d/%d\n", (int)GetMessageTime(),`, are timestamps ever different for VK_CONTROL and VK_MENU on AltGr?
* Does it help to disable the 2.8.0 patch with option `OldAltGrDetection=true`?

**Comment by mintty** [comment]
I've tested this myself with TeamViewer and could confirm my suspicion:
For AltGr, TeamViewer generate VK_CONTROL and VK_MENU events with different timestamps, most of the time. Sometimes the timestamps are equal and in those cases, the expected character is produced.
This is a deviation from Microsoft behaviour which is to generate both events with the same timestamp for an AltGr key. Please report that as a bug to TeamViewer.

**Comment by danowar2k** [comment]
Okay, here I am again. Regarding the first comment, do I still need to do all of these things? It seems you already did some or all of them?

I'll try to file a Teamviewer bug, I've been searching for their issue tracker anyway...

**Comment by danowar2k** [comment]
Okay, I've opened a [thread on the Teamviewer community forum](https://community.teamviewer.com/t5/TeamViewer-13/Teamviewer-13-1-3629-Can-t-use-Alt-gr-special-characters-on/m-p/38717#M2019).

**Comment by danowar2k** [comment]
> With the version you tested, does AltGr+~ not enter anything anymore?

Whoops, I've corrected the above report, That was a copy/paste error.

**Comment by mintty** [comment]
OK, sorry for not relating to the "p" or "k" earlier; this is because Ctrl+Alt+some-keys generates an escape sequence, the beginning of which you do not see in the shell command line.
Also Ctrl+Alt+key combinations are typically available for control purposes in terminal environments, they may invoke certain functions in editors, open menus and all sorts of things. That's why they need to remain separated from AltGr, and they need to be recognizable as distinct from AltGr. I could implement a threshhold as a workaround, so if the sequence of VK_CONTROL and VK_MENU is really fast, it will count as AltGr. The design issue of course is, how to define the threshhold. And it would not be the default as there is a risk of misinterpretation and it's only a workaround for a third-party bug.
Thanks for posting to their forum.

**Comment by danowar2k** [comment]
OK, some ideas and questions...

Can you check at the beginning (when opening a terminal) or when a key is typed if the used keyboard has either a right Alt key, an Alt gr key or both? If no right Alt key is present then you would know when Alt gr is typed, couldn't you?

How many keyboards exist that have a right Alt key anyway? I only know keyboards with Alt gr (but I am living a sheltered life) ;-)

If using TeamViewer with other terminals/applications works, how do they do it? Like, how does IntelliJ does it? (https://github.com/JetBrains/intellij-community/tree/master/plugins/terminal)
Or putty (https://github.com/github/putty)?
Ord cmder (https://github.com/cmderdev/cmder)? (Which I didn't know existed until a few minutes ago)

Or is this because of some necessary layer because of Cygwin/Unix like libraries?

**Comment by mintty** [comment]
> Can you check at the beginning (when opening a terminal) or when a key is typed if the used keyboard has either a right Alt key, an Alt gr key or both?

I don't think that's possible at the Windows API.

> If no right Alt key is present then you would know when Alt gr is typed, couldn't you?

But we could have an optional workaround based on this assumption.

And no, I wasn't referring to the existing option "Ctrl+left-Alt is AltGr" as a workaround (as I was simply not aware of this old option...).

> If using TeamViewer with other terminals/applications works, how do they do it?
> ...
> Or is this because of some necessary layer because of Cygwin/Unix like libraries?

I don't know, see my previous question, please: In those applications, does Ctrl+Alt+~ work like AltGr? That would mean they apply a simplified AltGr detection (just like mintty with the option just discussed) but at the cost of not being able to distinguish true Ctrl+Alt anymore. Please test this.

> How about an option ... where you can manually set the timestamp? Or whether a right Alt should be recognized as Alt gr automatically?

That would be an optional workaround, as mentioned above. Note however, that the timestamp difference observed from TeamViewer is 16ms while pressing Ctrl+Alt manually can yield a timestamp difference of also 16ms or even 15ms.

Are you/your company a paying user of TeamViewer? In that case, you may file a support ticket, please, to raise more attention than a forum posting. Unfortunately, they are so ignorant as to not allow bug reports from non-paying users.

**Comment by mintty** [comment]
>  Any pointers which are some typical Ctrl+Alt combinations used in terminals / Unix applications?

My own editor mined uses some. But for our purpose it is sufficient to check if the combination does not produce € etc like AltGr does.

**Comment by danowar2k** [comment]
Well, I can't find any Ctrl+Alt hotkeys for the IntelliJ terminal or SSH terminal. Ctrl+Left Alt inside them produces Alt gr characters. I can't set any hotkeys for them that use Ctrl+Left Alt+Key for any special function.

cmder has a hotkey Ctrl+Left Alt+u, which works as intended via TeamViewer. As there is no Alt gr char for "u", I can't check for that. Other Alt gr combinations work as intended, e.g. @,{,},~.

cmder also has the ability to assign hotkeys to functions. I've assigned CTRL+ALT+q to "Show Status Bar". Now CTRL+LEFT ALT+q toggles the status bar, and doesn't produce "@" anymore. Alt gr+q still produces "@".

Keep in mind I'm using cmder as a Windows cmd replacement at this moment, I am not connected via SSH to a Linux machine and am not using a bash.

Wouldn't it be only sufficient if there was a function assigned to Crtl+Left Alt+KEY AND a key like "@" assigned to Alt gr+KEY?
Because the program could check and/or forbid combinations of Ctrl+Left Alt+KEY internally which have a Alt gr+Key character output, couldn't it? Or it could check internally if the pressed combo is used for a function and only if it isn't it would default to Alt gr behaviour.

**Comment by mintty** [comment]
Let's sort this out; explicit Ctrl+Alt hotkeys are not really interesting in this context.

> ... IntelliJ terminal or SSH terminal. Ctrl+Left Alt inside them produces Alt gr characters.
> cmder ... Now CTRL+LEFT ALT+q ... doesn't produce "@" anymore.

This confirms that IntelliJ, SSH terminal (whichever) and cmder do not distinguish between Ctrl+Alt and AltGr. If you override Ctrl+Alt, you override the respective AltGr as well. That's not the idea of AltGr, and we want a more flexible handling, as you say yourself:

> Wouldn't it be only sufficient if there was a function assigned to Crtl+Left Alt+KEY AND a key like "@" assigned to Alt gr+KEY?

Sure, but for that reason we cannot apply the simplified AltGr detection approach that is apparently taken by cmder etc (except with option "Ctrl+left-Alt is AltGr").

> Because the program could check and/or forbid combinations of Ctrl+Left Alt+KEY internally which have a Alt gr+Key character output, couldn't it?

We discussed this already. I do not know any Windows API function that would easily reveal this information. And if there were one, this approach would make key handling quite inconsistent.

> Or it could check internally if the pressed combo is used for a function and only if it isn't it would default to Alt gr behaviour.

A terminal is there to process those key combinations and generate the respective characters or escape sequences to pass to a terminal application (e.g. an editor). The terminal cannot know in what way those are used by the application.

**Comment by mintty** [comment]
The checkbox option is about left-Alt, and its finer-tuned variant would be to "recognize left-Alt as AltGr only if it comes about the same time".
The other possible new option would be to recognize right-Alt as AltGr (maybe under the same condition) and this one could perhaps even be the default, under the keyboard considerations you suggested.

**Comment by mintty** [comment]
Wait, we have AltGr being derived from right-Alt already. There's something else that keeps it from becoming effective. And it's not the 2.8.0 patch, it's the section after it, where right-Alt with a different timestamp than the preceding Control is marked is having Control too. What about this (not yet tested with TeamViewer):
```
  bool altgr = ralt | ctrl_lalt_altgr;
  if (altgr)
    ctrl = rctrl;

  mod_keys mods = shift * MDK_SHIFT | alt * MDK_ALT | ctrl * MDK_CTRL;
```

**Comment by danowar2k** [comment]
I cloned mintty into my IntelliJ and discovered that I need a separate CLion IDE and IntelliJ doesn't have a C development plugin yet...

Anyway, I now scanned through the whole mininput.c file. I was a bit confused at first because normally I expect constants and macro definitions at the beginning of a file and not at the moment they are first needed.
(I remembered C doesn't have classes while I was constantly looking for a class definition or instinctively trying to group methods together in classes)

Ok, so the function you mean is probably `win_key_down`...I'm having trouble with some variables (damn IDE), I'm confused what"extended" means...
`  bool extended = HIWORD(lp) & KF_EXTENDED;
`
An extended keypress?

Anyway, I found the line. It's when all the modifiers are detected and stored in bool variables.

So if an AltGr press is detected, Ctrl is only pressed if the RCtrl is pressed? Won't that mess things up?
I'll compile...

**Comment by danowar2k** [comment]
Keys: AltGr + {
Result: {
Logged:
```
win_key_down 0011 VK_CONTROL scan 29
 Shift 0:0/0 Ctrl 1:1/0 Alt 0:0/0
win_key_down 0012 VK_MENU scan 312
 Shift 0:0/0 Ctrl 1:1/0 Alt 1:0/1
win_key_down 0037 37 scan 8
 Shift 0:0/0 Ctrl 1:1/0 Alt 1:0/1
 <-0... char_key
  win_key_up 0037 37
  win_key_up 0011 VK_CONTROL
  win_key_up 0011 VK_CONTROL
  win_key_up 0012 VK_MENU
```

Keys: AltGr + ~
Result: ~
Logged:
```
win_key_down 0011 VK_CONTROL scan 29
 Shift 0:0/0 Ctrl 1:1/0 Alt 0:0/0
win_key_down 0012 VK_MENU scan 312
 Shift 0:0/0 Ctrl 1:1/0 Alt 1:0/1
win_key_down 00BB VK_OEM_PLUS scan 27
 Shift 0:0/0 Ctrl 1:1/0 Alt 1:0/1
 <-0... char_key
  win_key_up 00BB VK_OEM_PLUS
  win_key_up 0011 VK_CONTROL
  win_key_up 0011 VK_CONTROL
  win_key_up 0012 VK_MENU
```

This seems to do the trick...but are there never things where LCtrl+AltGr does something?
This fix would prevent that, wouldn't it?

**Comment by mintty** [comment]
> So if an AltGr press is detected, Ctrl is only pressed if the RCtrl is pressed? Won't that mess things up?

The idea of this, let's say diagnostic, patch was to check whether AltGr detection was confused by the determined Control state which would trigger other handling with priority. And that is apparently the case.

> things where LCtrl+AltGr does something?

Yes, there are. E.g. Ctrl+@ should be the NUL character even if @ is AltGr+q.

> This fix would prevent that, wouldn't it?

Not completely, you could still use the right-Ctrl key if you have it :)
Or press AltGr first, then left-Ctrl.
But in fact, "normal" left-Ctrl+AltGr should not be affected by default.
Therefore I propose an option `AltGrThreshhold` to define the acceptable delay of Ctrl and right-Alt events to be interpreted as AltGr, with a suggestion to set it to 16 or 20 for TeamViewer.

**Comment by danowar2k** [comment]
> Therefore I propose

I'm unsure at the moment if you want/need any confirmation or comment from me...The quote sounds to me like you're asking for it, but at the same time this is your project...

So: I personally would rather have an option "Always interpret RAlt as AltGr" but I'd also be content with your variant.

**Comment by mintty** [comment]
Please test the uploaded workaround.

**Comment by danowar2k** [comment]
I've seen it, just haven't got the time yet...

**Comment by danowar2k** [comment]
Testing right now...

**Comment by danowar2k** [comment]
I explicitly downloaded the commit with the workaround: https://github.com/mintty/mintty/tree/9b68e247d243ac0e3791476fa774264f8778c22b .

I compiled it and started it, but didn't find any option to set the delay.
Shouldn't it be in Options -> Keys?

**Comment by danowar2k** [comment]
Ok, I started the terminal with ` ./mintty.exe -o CtrlAltDelayAltGr=16`.

AltGr is recognized 70% of the times, it seems. Sometimes it isn't recognized and I have to try again and again (let go of AltGr and press it again) and finally it is recognized.

**Comment by mintty** [comment]
> EDIT: ./mintty.exe -o CtrlAltDelayAltGr=50 seems to do the trick. Is that still acceptable?

Thanks for testing, I will change the recommendation in the manual and wiki page accordingly.
Whether it's acceptable is a user question. It means that non-AltGr Ctrl+Alt combinations are not recognized if Ctrl and Alt are hit to quickly. That's also the reason it cannot be the default.

Unacceptable is the ignorance of TeamViewer to recognize a bug report. I informed them at their sales channel which is the only contact they provide, not raising much interest though... Shame on such companies.

**Comment by BrianInglis** [comment]
Emails to sales contacts go to salesmen, who in most companies are only interested in orders from their area(s) for their product(s); if I ask a question about a feature or a new version, and don't get a response, I assume the product will not do what I require. 
Many of these commercial software vendors require you to have a support contract, and often be a registered contact, to get access to their system, before you can submit a bug report or download patches, to reduce any burden from those who have not paid, or are not currently paying, for the privilege. 
Some (many?) product users may have free personal use licences without bug reporting rights, "supported" by other users in community forums: it's up to product users to report product issues via available channels for both internal and external software support.

**Comment by danowar2k** [comment]
Yeah, I was looking for someone who has a paid Teamviewer license to help with that, no luck yet....

**Comment by mintty** [comment]
@BrianInglis, I am aware of this policy which they apply, and I understand it to a certain extent.
But I am not a personal free user and not a user at all, rather responsible for an *impacted* third-party product, from their point of view; I think their should be a channel for those.
Yeah, we'll have to hope a licensee who is also cygwin user will step in...

**Comment by mintty** [comment]
Released 2.9.1.
